From 2ecbfaf1f1f79cb2f318bf474f2195b66e1c9bd7 Mon Sep 17 00:00:00 2001
From: Armin Novak <armin.novak@thincast.com>
Date: Wed, 30 Sep 2026 11:46:17 +0200
Subject: [PATCH] [winpr,clipboard] Add ClipboardRegisterSynthesizerEx

Refactor clipboard internal handling. ClipboardRegisterSynthesizer does
call the synthesizer with the format of the data in the clipboard, which
is also accessible from wClipboard but fails to pass the synthesized
format ID. This limits usefulness of the callback as there needs to be
one for each target format.

ClipboardRegisterSynthesizerEx calls the synthesizer with the format to
synthesize, so now you got source and destination formats available and
can convert them however desired.
---
 winpr/libwinpr/clipboard/clipboard.c          |  105 +-
 winpr/libwinpr/clipboard/clipboard.h          |   20 +
 winpr/libwinpr/clipboard/synthetic.c          | 1010 ++++++-----------
 winpr/libwinpr/clipboard/synthetic_file.c     |  140 ++-
 .../clipboard/test/TestClipboardFormats.c     |  245 +++-
 winpr/libwinpr/utils/image.c                  |   53 +-
 6 files changed, 792 insertions(+), 781 deletions(-)

diff --git a/winpr/libwinpr/clipboard/clipboard.c b/winpr/libwinpr/clipboard/clipboard.c
index f0ffd9defe8a..5c36869c77b5 100644
--- a/winpr/libwinpr/clipboard/clipboard.c
+++ b/winpr/libwinpr/clipboard/clipboard.c
@@ -34,6 +34,7 @@
 
 const char* const mime_text_plain = "text/plain";
 const char* const mime_text_utf8 = "text/plain;charset=utf-8";
+const char* const mime_text_UTF8_STRING = "UTF8_STRING";
 
 /**
  * Clipboard (Windows):
@@ -255,47 +256,96 @@ UINT32 ClipboardRegisterFormat(wClipboard* clipboard, const char* name)
 	return format->formatId;
 }
 
-BOOL ClipboardRegisterSynthesizer(wClipboard* clipboard, UINT32 formatId, UINT32 syntheticId,
-                                  CLIPBOARD_SYNTHESIZE_FN pfnSynthesize)
+WINPR_ATTR_NODISCARD
+static wClipboardSynthesizer* registerSynthesizer(wClipboard* clipboard, UINT32 formatId,
+                                                  UINT32 syntheticId)
 {
-	UINT32 index = 0;
-	wClipboardFormat* format = nullptr;
-	wClipboardSynthesizer* synthesizer = nullptr;
-
 	if (!clipboard)
-		return FALSE;
+	{
+		WLog_WARN(TAG,
+		          "Failed to register synthesizer for %s[0x%04" PRIx32 "] to %s [0x%04" PRIx32
+		          "]: clipboard=NULL",
+		          ClipboardGetFormatName(clipboard, formatId), formatId,
+		          ClipboardGetFormatName(clipboard, syntheticId), syntheticId);
+		return nullptr;
+	}
 
-	format = ClipboardFindFormat(clipboard, formatId, nullptr);
+	WLog_DBG(TAG, "Registering synthesizer for %s[0x%04" PRIx32 "] to %s [0x%04" PRIx32 "]",
+	         ClipboardGetFormatName(clipboard, formatId), formatId,
+	         ClipboardGetFormatName(clipboard, syntheticId), syntheticId);
+
+	wClipboardFormat* format = ClipboardFindFormat(clipboard, formatId, nullptr);
 
 	if (!format)
-		return FALSE;
+	{
+		WLog_WARN(TAG,
+		          "Failed to register synthesizer for %s[0x%04" PRIx32 "] to %s [0x%04" PRIx32
+		          "]: format=NULL",
+		          ClipboardGetFormatName(clipboard, formatId), formatId,
+		          ClipboardGetFormatName(clipboard, syntheticId), syntheticId);
+		return nullptr;
+	}
 
 	if (format->formatId == syntheticId)
-		return FALSE;
+	{
+		WLog_WARN(TAG,
+		          "Failed to register synthesizer for %s[0x%04" PRIx32 "] to %s [0x%04" PRIx32
+		          "]: formats equal",
+		          ClipboardGetFormatName(clipboard, formatId), formatId,
+		          ClipboardGetFormatName(clipboard, syntheticId), syntheticId);
+		return nullptr;
+	}
 
-	synthesizer = ClipboardFindSynthesizer(format, formatId);
+	wClipboardSynthesizer* synthesizer = ClipboardFindSynthesizer(format, formatId);
 
 	if (!synthesizer)
 	{
-		wClipboardSynthesizer* tmpSynthesizer = nullptr;
 		UINT32 numSynthesizers = format->numSynthesizers + 1;
-		tmpSynthesizer = (wClipboardSynthesizer*)realloc(
+		wClipboardSynthesizer* tmpSynthesizer = (wClipboardSynthesizer*)realloc(
 		    format->synthesizers, numSynthesizers * sizeof(wClipboardSynthesizer));
 
 		if (!tmpSynthesizer)
-			return FALSE;
+		{
+			WLog_WARN(TAG,
+			          "Failed to register synthesizer for %s[0x%04" PRIx32 "] to %s [0x%04" PRIx32
+			          "]: failed to allocate new synthesizer",
+			          ClipboardGetFormatName(clipboard, formatId), formatId,
+			          ClipboardGetFormatName(clipboard, syntheticId), syntheticId);
+			return nullptr;
+		}
 
 		format->synthesizers = tmpSynthesizer;
 		format->numSynthesizers = numSynthesizers;
-		index = numSynthesizers - 1;
+		const UINT32 index = numSynthesizers - 1;
 		synthesizer = &(format->synthesizers[index]);
 	}
+	return synthesizer;
+}
+
+BOOL ClipboardRegisterSynthesizer(wClipboard* clipboard, UINT32 formatId, UINT32 syntheticId,
+                                  CLIPBOARD_SYNTHESIZE_FN pfnSynthesize)
+{
+	wClipboardSynthesizer* synthesizer = registerSynthesizer(clipboard, formatId, syntheticId);
+	if (!synthesizer)
+		return FALSE;
 
 	synthesizer->syntheticId = syntheticId;
 	synthesizer->pfnSynthesize = pfnSynthesize;
 	return TRUE;
 }
 
+BOOL ClipboardRegisterSynthesizerEx(wClipboard* clipboard, UINT32 formatId, UINT32 syntheticId,
+                                    CLIPBOARD_SYNTHESIZE_FN pfnSynthesize)
+{
+	wClipboardSynthesizer* synthesizer = registerSynthesizer(clipboard, formatId, syntheticId);
+	if (!synthesizer)
+		return FALSE;
+
+	synthesizer->syntheticId = syntheticId;
+	synthesizer->pfnSynthesizeEx = pfnSynthesize;
+	return TRUE;
+}
+
 UINT32 ClipboardCountFormats(wClipboard* clipboard)
 {
 	UINT32 count = 0;
@@ -430,12 +480,7 @@ const char* ClipboardGetFormatName(wClipboard* clipboard, UINT32 formatId)
 
 void* ClipboardGetData(wClipboard* clipboard, UINT32 formatId, UINT32* pSize)
 {
-	UINT32 SrcSize = 0;
-	UINT32 DstSize = 0;
-	void* pSrcData = nullptr;
 	void* pDstData = nullptr;
-	wClipboardFormat* format = nullptr;
-	wClipboardSynthesizer* synthesizer = nullptr;
 
 	if (!clipboard || !pSize)
 	{
@@ -446,7 +491,7 @@ void* ClipboardGetData(wClipboard* clipboard, UINT32 formatId, UINT32* pSize)
 	}
 
 	*pSize = 0;
-	format = ClipboardFindFormat(clipboard, clipboard->formatId, nullptr);
+	wClipboardFormat* format = ClipboardFindFormat(clipboard, clipboard->formatId, nullptr);
 
 	if (!format)
 	{
@@ -454,12 +499,12 @@ void* ClipboardGetData(wClipboard* clipboard, UINT32 formatId, UINT32* pSize)
 		return nullptr;
 	}
 
-	SrcSize = clipboard->size;
-	pSrcData = clipboard->data;
+	const UINT32 SrcSize = clipboard->size;
+	void* pSrcData = clipboard->data;
 
 	if (formatId == format->formatId)
 	{
-		DstSize = SrcSize;
+		UINT32 DstSize = SrcSize;
 		pDstData = malloc(DstSize);
 
 		if (!pDstData)
@@ -470,9 +515,9 @@ void* ClipboardGetData(wClipboard* clipboard, UINT32 formatId, UINT32* pSize)
 	}
 	else
 	{
-		synthesizer = ClipboardFindSynthesizer(format, formatId);
+		wClipboardSynthesizer* synthesizer = ClipboardFindSynthesizer(format, formatId);
 
-		if (!synthesizer || !synthesizer->pfnSynthesize)
+		if (!synthesizer || !(synthesizer->pfnSynthesize || synthesizer->pfnSynthesizeEx))
 		{
 			WLog_ERR(TAG, "No synthesizer for format %s [0x%08" PRIx32 "] --> %s [0x%08" PRIx32 "]",
 			         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId,
@@ -480,8 +525,12 @@ void* ClipboardGetData(wClipboard* clipboard, UINT32 formatId, UINT32* pSize)
 			return nullptr;
 		}
 
-		DstSize = SrcSize;
-		pDstData = synthesizer->pfnSynthesize(clipboard, format->formatId, pSrcData, &DstSize);
+		UINT32 DstSize = SrcSize;
+		if (synthesizer->pfnSynthesizeEx)
+			pDstData = synthesizer->pfnSynthesizeEx(clipboard, synthesizer->syntheticId, pSrcData,
+			                                        &DstSize);
+		else
+			pDstData = synthesizer->pfnSynthesize(clipboard, format->formatId, pSrcData, &DstSize);
 		if (pDstData)
 			*pSize = DstSize;
 	}
diff --git a/winpr/libwinpr/clipboard/clipboard.h b/winpr/libwinpr/clipboard/clipboard.h
index 3946ee480a50..c642bb0ce735 100644
--- a/winpr/libwinpr/clipboard/clipboard.h
+++ b/winpr/libwinpr/clipboard/clipboard.h
@@ -29,6 +29,7 @@ typedef struct
 {
 	UINT32 syntheticId;
 	CLIPBOARD_SYNTHESIZE_FN pfnSynthesize;
+	CLIPBOARD_SYNTHESIZE_FN pfnSynthesizeEx;
 } wClipboardSynthesizer;
 
 typedef struct
@@ -74,5 +75,24 @@ WINPR_LOCAL WINPR_ATTR_NODISCARD char* parse_uri_to_local_file(const char* uri,
 
 extern const char* const mime_text_plain;
 extern const char* const mime_text_utf8;
+extern const char* const mime_text_UTF8_STRING;
+
+/** @brief Same as \ref ClipboardRegisterSynthesizer but with different \ref pfnSynthesize calling
+ * convention. While functions registered with \ref ClipboardRegisterSynthesizer will call \ref
+ * pfnSynthesize with the \ref formatId of the data in the clipboard functions registered with this
+ * will be called with \ref syntheticId instead.
+ *
+ *  @param clipboard The clipboard to use
+ *  @param formatId The format of the data in the clipboard
+ *  @param syntheticId The format of the data to synthesize
+ *  @param pfnSynthesize The function to call to synthesize
+ *  @return TRUE for success, FALSE otherwise
+ *
+ *  @since version 3.32.2
+ */
+WINPR_ATTR_NODISCARD
+WINPR_LOCAL BOOL ClipboardRegisterSynthesizerEx(wClipboard* clipboard, UINT32 formatId,
+                                                UINT32 syntheticId,
+                                                CLIPBOARD_SYNTHESIZE_FN pfnSynthesize);
 
 #endif /* WINPR_CLIPBOARD_PRIVATE_H */
diff --git a/winpr/libwinpr/clipboard/synthetic.c b/winpr/libwinpr/clipboard/synthetic.c
index f736207252d4..504770bc0365 100644
--- a/winpr/libwinpr/clipboard/synthetic.c
+++ b/winpr/libwinpr/clipboard/synthetic.c
@@ -33,14 +33,34 @@
 
 static const char mime_html[] = "text/html";
 static const char mime_ms_html[] = "HTML Format";
-static const char* mime_bitmap[] = { "image/bmp", "image/x-bmp", "image/x-MS-bmp",
-	                                 "image/x-win-bitmap" };
 
+#define BITMAP_MIME_TYPES "image/bmp", "image/x-bmp", "image/x-MS-bmp", "image/x-win-bitmap"
+static const char* mime_bitmap[] = { BITMAP_MIME_TYPES };
+
+#if defined(WINPR_UTILS_IMAGE_WEBP)
 static const char mime_webp[] = "image/webp";
+#endif
+#if defined(WINPR_UTILS_IMAGE_PNG)
 static const char mime_png[] = "image/png";
+#endif
+#if defined(WINPR_UTILS_IMAGE_JPEG)
 static const char mime_jpeg[] = "image/jpeg";
+#endif
 static const char mime_tiff[] = "image/tiff";
 
+static const char* mime_images[] = {
+#if defined(WINPR_UTILS_IMAGE_WEBP)
+	mime_webp,
+#endif
+#if defined(WINPR_UTILS_IMAGE_PNG)
+	mime_png,
+#endif
+#if defined(WINPR_UTILS_IMAGE_JPEG)
+	mime_jpeg,
+#endif
+	mime_tiff, BITMAP_MIME_TYPES
+};
+
 static const BYTE enc_base64url[] =
     "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
 
@@ -117,91 +137,6 @@ static inline char* b64_encode(const BYTE* WINPR_RESTRICT data, size_t length, s
  * http://msdn.microsoft.com/en-us/library/windows/desktop/ff729168/
  */
 
-/**
- * "CF_TEXT":
- *
- * Null-terminated ANSI text with CR/LF line endings.
- */
-
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_cf_text(wClipboard* clipboard, UINT32 formatId, const void* data,
-                                          UINT32* pSize)
-{
-	size_t size = 0;
-	char* pDstData = nullptr;
-
-	if (formatId == CF_UNICODETEXT)
-	{
-		size_t strsize = 0;
-		char* str = ConvertWCharNToUtf8Alloc(data, *pSize / sizeof(WCHAR), &strsize);
-
-		if (!str || (strsize > UINT32_MAX))
-		{
-			winpr_znfree(str, strsize);
-			return nullptr;
-		}
-
-		size = strsize;
-		pDstData = ConvertLineEndingToCRLF(str, &size);
-		winpr_znfree(str, strsize);
-		*pSize = (UINT32)size;
-		return pDstData;
-	}
-	else if (formatId == ClipboardGetFormatId(clipboard, mime_text_utf8))
-	{
-		size = *pSize;
-		pDstData = ConvertLineEndingToCRLF(data, &size);
-
-		if (!pDstData || (size == 0) || (size >= UINT32_MAX))
-		{
-			winpr_znfree(pDstData, size);
-			return nullptr;
-		}
-
-		size_t escsize = 0;
-		char* escaped = winpr_utf8ToUtfEscapedString(pDstData, size, &escsize);
-		winpr_znfree(pDstData, size);
-		if (!escaped || (escsize == 0) || (escsize > UINT32_MAX))
-		{
-			winpr_znfree(escaped, escsize);
-			return nullptr;
-		}
-		/* CF_TEXT is null terminated, the size must include the terminator */
-		*pSize = (UINT32)escsize + 1;
-		return escaped;
-	}
-	else if ((formatId == CF_TEXT) || (formatId == CF_OEMTEXT) ||
-	         (formatId == ClipboardGetFormatId(clipboard, mime_text_plain)))
-	{
-		size = *pSize;
-		pDstData = ConvertLineEndingToCRLF(data, &size);
-
-		if (!pDstData || (size >= UINT32_MAX))
-		{
-			winpr_znfree(pDstData, size);
-			return nullptr;
-		}
-
-		/* CF_TEXT is null terminated, the size must include the terminator */
-		*pSize = (UINT32)size + 1;
-		return pDstData;
-	}
-
-	return nullptr;
-}
-
-/**
- * "CF_OEMTEXT":
- *
- * Null-terminated OEM text with CR/LF line endings.
- */
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_cf_oemtext(wClipboard* clipboard, UINT32 formatId,
-                                             const void* data, UINT32* pSize)
-{
-	return clipboard_synthesize_cf_text(clipboard, formatId, data, pSize);
-}
-
 /**
  * "CF_LOCALE":
  *
@@ -209,85 +144,49 @@ static void* clipboard_synthesize_cf_oemtext(wClipboard* clipboard, UINT32 forma
  */
 WINPR_ATTR_MALLOC(free, 1)
 static void* clipboard_synthesize_cf_locale(WINPR_ATTR_UNUSED wClipboard* clipboard,
-                                            WINPR_ATTR_UNUSED UINT32 formatId,
-                                            WINPR_ATTR_UNUSED const void* data,
+                                            UINT32 dstFormatId, WINPR_ATTR_UNUSED const void* data,
                                             WINPR_ATTR_UNUSED UINT32* pSize)
 {
-	UINT32* pDstData = nullptr;
-	pDstData = (UINT32*)malloc(sizeof(UINT32));
-
-	if (!pDstData)
-		return nullptr;
-
-	*pDstData = 0x0409; /* English - United States */
-	return (void*)pDstData;
-}
-
-/**
- * "CF_UNICODETEXT":
- *
- * Null-terminated UTF-16 text with CR/LF line endings.
- */
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_cf_unicodetext(wClipboard* clipboard, UINT32 formatId,
-                                                 const void* data, UINT32* pSize)
-{
-	size_t size = 0;
-	char* crlfStr = nullptr;
-	WCHAR* pDstData = nullptr;
-
-	WINPR_ASSERT(clipboard);
-	if (!pSize || (*pSize > INT32_MAX))
-		return nullptr;
-	if (clipboard->formatId != CF_UNICODETEXT)
+	if (dstFormatId != CF_LOCALE)
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, dstFormatId), dstFormatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
 		return nullptr;
+	}
 
-	if (formatId == ClipboardGetFormatId(clipboard, mime_text_utf8))
+	switch (clipboard->formatId)
 	{
-		size_t len = 0;
-
-		size = *pSize;
-		crlfStr = ConvertLineEndingToCRLF((const char*)data, &size);
-
-		if (!crlfStr)
-			return nullptr;
-
-		pDstData = ConvertUtf8NToWCharAlloc(crlfStr, size, &len);
-		free(crlfStr);
-
-		if ((len < 1) || ((len + 1) > UINT32_MAX / sizeof(WCHAR)))
+		case CF_TEXT:
+		case CF_UNICODETEXT:
+		case CF_OEMTEXT:
+			break;
+		default:
 		{
-			free(pDstData);
-			return nullptr;
+			const UINT32 formatIdText = ClipboardRegisterFormat(clipboard, mime_text_plain);
+			const UINT32 formatIdUtf = ClipboardRegisterFormat(clipboard, mime_text_plain);
+			if ((clipboard->formatId != formatIdText) && (clipboard->formatId != formatIdUtf))
+			{
+				WLog_ERR(TAG,
+				         "Unuspported source format %s [0x%04" PRIx32
+				         "], trying to convert to %s [0x%04" PRIx32 "]",
+				         ClipboardGetFormatName(clipboard, clipboard->formatId),
+				         clipboard->formatId, ClipboardGetFormatName(clipboard, dstFormatId),
+				         dstFormatId);
+				return nullptr;
+			}
 		}
-
-		const size_t slen = (len + 1) * sizeof(WCHAR);
-		*pSize = (UINT32)slen;
+		break;
 	}
-	if ((formatId == CF_TEXT) || (formatId == CF_OEMTEXT) ||
-	    (formatId == ClipboardGetFormatId(clipboard, mime_text_plain)))
-	{
-		size_t len = 0;
 
-		size = *pSize;
-		crlfStr = ConvertLineEndingToCRLF((const char*)data, &size);
-
-		if (!crlfStr)
-			return nullptr;
-
-		pDstData = ConvertUtf8NToWCharAlloc(crlfStr, size, &len);
-		free(crlfStr);
-
-		if ((len < 1) || ((len + 1) > UINT32_MAX / sizeof(WCHAR)))
-		{
-			free(pDstData);
-			return nullptr;
-		}
+	UINT32* pDstData = (UINT32*)calloc(1, sizeof(UINT32));
 
-		const size_t slen = (len + 1) * sizeof(WCHAR);
-		*pSize = (UINT32)slen;
-	}
+	if (!pDstData)
+		return nullptr;
 
+	*pDstData = 0x0409; /* English - United States */
 	return (void*)pDstData;
 }
 
@@ -297,15 +196,15 @@ static void* clipboard_synthesize_cf_unicodetext(wClipboard* clipboard, UINT32 f
  * Null-terminated UTF-8 string with LF line endings.
  */
 WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_utf8_string(wClipboard* clipboard, UINT32 formatId,
-                                              const void* data, UINT32* pSize)
+static void* clipboard_synthesize_string(wClipboard* clipboard, UINT32 dstFormatId,
+                                         const void* data, UINT32* pSize)
 {
 	WINPR_ASSERT(clipboard);
 
 	// Step 1: convert to utf-8
 	char* utf8 = nullptr;
 	size_t utf8len = 0;
-	if (formatId == CF_UNICODETEXT)
+	if (clipboard->formatId == CF_UNICODETEXT)
 	{
 		size_t size = 0;
 		utf8 = ConvertWCharNToUtf8Alloc(data, *pSize / sizeof(WCHAR), &size);
@@ -316,8 +215,8 @@ static void* clipboard_synthesize_utf8_string(wClipboard* clipboard, UINT32 form
 		const size_t rc = ConvertLineEndingToLF(utf8, size);
 		utf8len = rc;
 	}
-	else if ((formatId == CF_TEXT) || (formatId == CF_OEMTEXT) ||
-	         (formatId == ClipboardGetFormatId(clipboard, mime_text_plain)))
+	else if ((clipboard->formatId == CF_TEXT) || (clipboard->formatId == CF_OEMTEXT) ||
+	         (clipboard->formatId == ClipboardGetFormatId(clipboard, mime_text_plain)))
 	{
 		const size_t size = *pSize;
 		utf8 = calloc(size + 1, sizeof(char));
@@ -335,7 +234,8 @@ static void* clipboard_synthesize_utf8_string(wClipboard* clipboard, UINT32 form
 		}
 		utf8len = (size_t)res;
 	}
-	else if (formatId == ClipboardGetFormatId(clipboard, mime_text_utf8))
+	else if ((clipboard->formatId == ClipboardGetFormatId(clipboard, mime_text_utf8)) ||
+	         (clipboard->formatId == ClipboardGetFormatId(clipboard, mime_text_UTF8_STRING)))
 	{
 		const size_t size = *pSize;
 		utf8 = strndup(data, size);
@@ -345,23 +245,34 @@ static void* clipboard_synthesize_utf8_string(wClipboard* clipboard, UINT32 form
 	}
 	else
 	{
-		WLog_ERR(TAG, "Unuspported source format %s [0x%04" PRIx32 "]",
-		         ClipboardGetFormatName(clipboard, formatId), formatId);
+		WLog_ERR(TAG,
+		         "Unuspported source format %s [0x%04" PRIx32
+		         "], trying to convert to %s [0x%04" PRIx32 "]",
+
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId,
+		         ClipboardGetFormatName(clipboard, dstFormatId), dstFormatId);
 		return nullptr;
 	}
 
-	switch (clipboard->formatId)
+	switch (dstFormatId)
 	{
 		case CF_UNICODETEXT:
 		{
-			size_t wlen = 0;
-			WCHAR* wstr = ConvertUtf8NToWCharAlloc(utf8, utf8len, &wlen);
+			size_t crlfLen = utf8len;
+			char* crlf = ConvertLineEndingToCRLF(utf8, &crlfLen);
 			winpr_znfree(utf8, utf8len);
+			if (!crlf)
+				return nullptr;
+
+			size_t wlen = 0;
+			WCHAR* wstr = ConvertUtf8NToWCharAlloc(crlf, crlfLen, &wlen);
+			winpr_znfree(crlf, crlfLen);
 			if (!wstr || (wlen == 0) || (wlen > UINT32_MAX / sizeof(WCHAR)))
 			{
 				winpr_znfree(wstr, wlen * sizeof(WCHAR));
 				return nullptr;
 			}
+
 			*pSize = WINPR_ASSERTING_INT_CAST(UINT32, wlen * sizeof(WCHAR));
 			return wstr;
 		}
@@ -380,7 +291,8 @@ static void* clipboard_synthesize_utf8_string(wClipboard* clipboard, UINT32 form
 			return escaped;
 		}
 		default:
-			if (clipboard->formatId == ClipboardGetFormatId(clipboard, mime_text_utf8))
+			if ((dstFormatId == ClipboardGetFormatId(clipboard, mime_text_utf8)) ||
+			    (dstFormatId == ClipboardGetFormatId(clipboard, mime_text_UTF8_STRING)))
 			{
 				if (utf8len > UINT32_MAX)
 				{
@@ -391,7 +303,7 @@ static void* clipboard_synthesize_utf8_string(wClipboard* clipboard, UINT32 form
 				return utf8;
 			}
 
-			if (clipboard->formatId == ClipboardGetFormatId(clipboard, mime_text_plain))
+			if (dstFormatId == ClipboardGetFormatId(clipboard, mime_text_plain))
 			{
 				size_t esclen = 0;
 				char* escaped = winpr_utf8ToUtfEscapedString(utf8, utf8len, &esclen);
@@ -405,128 +317,16 @@ static void* clipboard_synthesize_utf8_string(wClipboard* clipboard, UINT32 form
 				return escaped;
 			}
 
-			WLog_ERR(TAG, "Unuspported destination format %s [0x%04" PRIx32 "]",
+			WLog_ERR(TAG,
+			         "Unuspported destination format %s [0x%04" PRIx32
+			         "], trying to convert from %s [0x%04" PRIx32 "]",
+			         ClipboardGetFormatName(clipboard, dstFormatId), dstFormatId,
 			         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
 			winpr_znfree(utf8, utf8len);
 			return nullptr;
 	}
 }
 
-WINPR_ATTR_NODISCARD
-static BOOL is_format_bitmap(wClipboard* clipboard, UINT32 formatId)
-{
-	for (size_t x = 0; x < ARRAYSIZE(mime_bitmap); x++)
-	{
-		const char* mime = mime_bitmap[x];
-		const UINT32 altFormatId = ClipboardGetFormatId(clipboard, mime);
-		if (altFormatId == formatId)
-			return TRUE;
-	}
-
-	return FALSE;
-}
-
-/**
- * "CF_DIB":
- *
- * BITMAPINFO structure followed by the bitmap bits.
- */
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_cf_dib(wClipboard* clipboard, UINT32 formatId, const void* data,
-                                         UINT32* pSize)
-{
-	UINT32 SrcSize = 0;
-	UINT32 DstSize = 0;
-	BYTE* pDstData = nullptr;
-	SrcSize = *pSize;
-
-#if defined(WINPR_UTILS_IMAGE_DIBv5)
-	if (formatId == CF_DIBV5)
-	{
-		WLog_WARN(TAG, "[DIB] Unsupported destination format %s",
-		          ClipboardGetFormatName(clipboard, formatId));
-	}
-	else
-#endif
-	    if (is_format_bitmap(clipboard, formatId))
-	{
-		WINPR_BITMAP_FILE_HEADER pFileHeader = WINPR_C_ARRAY_INIT;
-		wStream sbuffer = WINPR_C_ARRAY_INIT;
-		wStream* s = Stream_StaticConstInit(&sbuffer, data, SrcSize);
-		if (!readBitmapFileHeader(s, &pFileHeader))
-			return nullptr;
-
-		DstSize = SrcSize - sizeof(BITMAPFILEHEADER);
-		pDstData = (BYTE*)malloc(DstSize);
-
-		if (!pDstData)
-			return nullptr;
-
-		data = (const void*)&((const BYTE*)data)[sizeof(BITMAPFILEHEADER)];
-		CopyMemory(pDstData, data, DstSize);
-		*pSize = DstSize;
-		return pDstData;
-	}
-	else
-	{
-		WLog_WARN(TAG, "[DIB] Unsupported destination format %s",
-		          ClipboardGetFormatName(clipboard, formatId));
-	}
-
-	return nullptr;
-}
-
-/**
- * "CF_DIBV5":
- *
- * BITMAPV5HEADER structure followed by the bitmap color space information and the bitmap bits.
- */
-#if defined(WINPR_UTILS_IMAGE_DIBv5)
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_cf_dibv5(wClipboard* clipboard, UINT32 formatId,
-                                           WINPR_ATTR_UNUSED const void* data,
-                                           WINPR_ATTR_UNUSED UINT32* pSize)
-{
-	if (formatId == CF_DIB)
-	{
-		WLog_WARN(TAG, "[DIBv5] Unsupported destination format %s",
-		          ClipboardGetFormatName(clipboard, formatId));
-	}
-	else if (is_format_bitmap(clipboard, formatId))
-	{
-		WLog_WARN(TAG, "[DIBv5] Unsupported destination format %s",
-		          ClipboardGetFormatName(clipboard, formatId));
-	}
-	else
-	{
-		BOOL handled = FALSE;
-#if defined(WINPR_UTILS_IMAGE_PNG)
-		{
-			const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_png);
-			if (formatId == altFormatId)
-			{
-			}
-		}
-#endif
-#if defined(WINPR_UTILS_IMAGE_JPEG)
-		{
-			const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_jpeg);
-			if (formatId == altFormatId)
-			{
-			}
-		}
-#endif
-		if (!handled)
-		{
-			WLog_WARN(TAG, "[DIBv5] Unsupported destination format %s",
-			          ClipboardGetFormatName(clipboard, formatId));
-		}
-	}
-
-	return nullptr;
-}
-#endif
-
 WINPR_ATTR_MALLOC(free, 1)
 static void* clipboard_prepend_bmp_header(const WINPR_BITMAP_INFO_HEADER* pInfoHeader,
                                           size_t offset, const void* data, size_t size,
@@ -591,11 +391,11 @@ static void* clipboard_prepend_bmp_header(const WINPR_BITMAP_INFO_HEADER* pInfoH
  */
 WINPR_ATTR_MALLOC(free, 1)
 static void* clipboard_synthesize_image_bmp(WINPR_ATTR_UNUSED wClipboard* clipboard,
-                                            UINT32 formatId, const void* data, UINT32* pSize)
+                                            UINT32 dstFormatId, const void* data, UINT32* pSize)
 {
 	UINT32 SrcSize = *pSize;
 
-	if (formatId == CF_DIB)
+	if (dstFormatId == CF_DIB)
 	{
 		if (SrcSize < sizeof(BITMAPINFOHEADER))
 			return nullptr;
@@ -610,37 +410,97 @@ static void* clipboard_synthesize_image_bmp(WINPR_ATTR_UNUSED wClipboard* clipbo
 		return clipboard_prepend_bmp_header(&header, offset, data, SrcSize, pSize);
 	}
 #if defined(WINPR_UTILS_IMAGE_DIBv5)
-	else if (formatId == CF_DIBV5)
+	else if (dstFormatId == CF_DIBV5)
 	{
-		WLog_WARN(TAG, "[BMP] Unsupported destination format %s",
-		          ClipboardGetFormatName(clipboard, formatId));
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, dstFormatId), dstFormatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
 	}
 #endif
 	else
 	{
-		WLog_WARN(TAG, "[BMP] Unsupported destination format %s",
-		          ClipboardGetFormatName(clipboard, formatId));
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, dstFormatId), dstFormatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
 	}
 
 	return nullptr;
 }
 
-#if defined(WINPR_UTILS_IMAGE_PNG) || defined(WINPR_UTILS_IMAGE_WEBP) || \
-    defined(WINPR_UTILS_IMAGE_JPEG)
+WINPR_ATTR_NODISCARD
+static BOOL format_is_image(wClipboard* clipboard, UINT32 formatId)
+{
+	switch (formatId)
+	{
+		case CF_DIB:
+		case CF_DIBV5:
+		case CF_TIFF:
+			return TRUE;
+		default:
+
+			for (size_t x = 0; x < ARRAYSIZE(mime_images); x++)
+			{
+				const char* mime = mime_images[x];
+				const UINT32 id = ClipboardRegisterFormat(clipboard, mime);
+				if (formatId == id)
+					return TRUE;
+			}
+			return FALSE;
+	}
+}
+
 WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_image_bmp_to_format(wClipboard* clipboard, UINT32 formatId,
-                                                      UINT32 bmpFormat, const void* data,
-                                                      UINT32* pSize)
+static void* clipboard_synthesize_image_dib_to_format(wClipboard* clipboard, UINT32 dstFormatId,
+                                                      const void* data, UINT32* pSize)
 {
 	WINPR_ASSERT(clipboard);
 	WINPR_ASSERT(data);
 	WINPR_ASSERT(pSize);
 
+	if ((clipboard->formatId != CF_DIB) && (clipboard->formatId != CF_DIBV5))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported source format %s [0x%04" PRIx32
+		         "], trying to convert to %s [0x%04" PRIx32 "]",
+
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId,
+		         ClipboardGetFormatName(clipboard, dstFormatId), dstFormatId);
+		return nullptr;
+	}
+	if (!format_is_image(clipboard, dstFormatId))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+
+		         ClipboardGetFormatName(clipboard, dstFormatId), dstFormatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+
+	UINT32 format = WINPR_IMAGE_BITMAP;
+#if defined(WINPR_UTILS_IMAGE_PNG)
+	if (dstFormatId == ClipboardRegisterFormat(clipboard, mime_png))
+		format = WINPR_IMAGE_PNG;
+#endif
+#if defined(WINPR_UTILS_IMAGE_JPEG)
+	if (dstFormatId == ClipboardRegisterFormat(clipboard, mime_jpeg))
+		format = WINPR_IMAGE_JPEG;
+#endif
+#if defined(WINPR_UTILS_IMAGE_WEBP)
+	if (dstFormatId == ClipboardRegisterFormat(clipboard, mime_webp))
+		format = WINPR_IMAGE_WEBP;
+#endif
+
 	size_t dsize = 0;
 	void* result = nullptr;
 
 	wImage* img = winpr_image_new();
-	void* bmp = clipboard_synthesize_image_bmp(clipboard, formatId, data, pSize);
+	void* bmp = clipboard_synthesize_image_bmp(clipboard, clipboard->formatId, data, pSize);
 	const UINT32 SrcSize = *pSize;
 	*pSize = 0;
 
@@ -650,7 +510,7 @@ static void* clipboard_synthesize_image_bmp_to_format(wClipboard* clipboard, UIN
 	if (winpr_image_read_buffer(img, bmp, SrcSize) <= 0)
 		goto fail;
 
-	result = winpr_image_write_buffer(img, bmpFormat, &dsize);
+	result = winpr_image_write_buffer(img, format, &dsize);
 	if (result)
 	{
 		if (dsize <= UINT32_MAX)
@@ -667,29 +527,65 @@ static void* clipboard_synthesize_image_bmp_to_format(wClipboard* clipboard, UIN
 	winpr_image_free(img, TRUE);
 	return result;
 }
-#endif
 
-#if defined(WINPR_UTILS_IMAGE_PNG)
+/**
+ * "CF_DIB":
+ *
+ * BITMAPINFO structure followed by the bitmap bits.
+ */
 WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_image_bmp_to_png(wClipboard* clipboard, UINT32 formatId,
-                                                   const void* data, UINT32* pSize)
+static void* clipboard_synthesize_cf_dib(const void* pvdata, size_t dataLen, UINT32* pSize)
 {
-	return clipboard_synthesize_image_bmp_to_format(clipboard, formatId, WINPR_IMAGE_PNG, data,
-	                                                pSize);
+	WINPR_ASSERT(pSize);
+	const BYTE* data = pvdata;
+
+	WINPR_BITMAP_FILE_HEADER pFileHeader = WINPR_C_ARRAY_INIT;
+	wStream sbuffer = WINPR_C_ARRAY_INIT;
+	wStream* s = Stream_StaticConstInit(&sbuffer, data, dataLen);
+	if (!readBitmapFileHeader(s, &pFileHeader))
+		return nullptr;
+
+	const size_t DstSize = dataLen - sizeof(BITMAPFILEHEADER);
+	BYTE* pDstData = (BYTE*)calloc(DstSize, 1);
+
+	if (!pDstData)
+		return nullptr;
+
+	const void* src = &data[sizeof(BITMAPFILEHEADER)];
+	memcpy(pDstData, src, DstSize);
+	*pSize = WINPR_ASSERTING_INT_CAST(UINT32, DstSize);
+	return pDstData;
 }
-#endif
 
-#if defined(WINPR_UTILS_IMAGE_PNG) || defined(WINPR_UTILS_IMAGE_WEBP) || \
-    defined(WINPR_UTILS_IMAGE_JPEG)
 WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_image_format_to_bmp(WINPR_ATTR_UNUSED wClipboard* clipboard,
-                                                      WINPR_ATTR_UNUSED UINT32 srcFormatId,
-                                                      const void* data, UINT32* pSize)
+static void* clipboard_synthesize_image_format_to_cf_dib(wClipboard* clipboard, UINT32 dstFormatId,
+                                                         const void* data, UINT32* pSize)
 {
 	WINPR_ASSERT(clipboard);
 	WINPR_ASSERT(data);
 	WINPR_ASSERT(pSize);
 
+	if ((dstFormatId != CF_DIB) && (dstFormatId != CF_DIBV5))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, dstFormatId), dstFormatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+
+	if (!format_is_image(clipboard, dstFormatId))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported source format %s [0x%04" PRIx32
+		         "], trying to convert to %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId,
+		         ClipboardGetFormatName(clipboard, dstFormatId), dstFormatId);
+		return nullptr;
+	}
+
+	void* result = nullptr;
 	BYTE* dst = nullptr;
 	const UINT32 SrcSize = *pSize;
 	size_t size = 0;
@@ -702,66 +598,16 @@ static void* clipboard_synthesize_image_format_to_bmp(WINPR_ATTR_UNUSED wClipboa
 		goto fail;
 
 	dst = winpr_image_write_buffer(image, WINPR_IMAGE_BITMAP, &size);
-	if ((size < sizeof(WINPR_BITMAP_FILE_HEADER)) || (size > UINT32_MAX))
-	{
-		free(dst);
-		dst = nullptr;
+	if ((size < sizeof(WINPR_BITMAP_FILE_HEADER)) || (size > UINT32_MAX) || !dst)
 		goto fail;
-	}
-	*pSize = (UINT32)size;
+
+	result = clipboard_synthesize_cf_dib(dst, size, pSize);
 
 fail:
 	winpr_image_free(image, TRUE);
-
-	if (dst)
-		memmove(dst, &dst[sizeof(WINPR_BITMAP_FILE_HEADER)],
-		        size - sizeof(WINPR_BITMAP_FILE_HEADER));
-	return dst;
-}
-#endif
-
-#if defined(WINPR_UTILS_IMAGE_PNG)
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_image_png_to_bmp(wClipboard* clipboard, UINT32 formatId,
-                                                   const void* data, UINT32* pSize)
-{
-	return clipboard_synthesize_image_format_to_bmp(clipboard, formatId, data, pSize);
-}
-#endif
-
-#if defined(WINPR_UTILS_IMAGE_WEBP)
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_image_bmp_to_webp(wClipboard* clipboard, UINT32 formatId,
-                                                    const void* data, UINT32* pSize)
-{
-	return clipboard_synthesize_image_bmp_to_format(clipboard, formatId, WINPR_IMAGE_WEBP, data,
-	                                                pSize);
-}
-
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_image_webp_to_bmp(wClipboard* clipboard, UINT32 formatId,
-                                                    const void* data, UINT32* pSize)
-{
-	return clipboard_synthesize_image_format_to_bmp(clipboard, formatId, data, pSize);
-}
-#endif
-
-#if defined(WINPR_UTILS_IMAGE_JPEG)
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_image_bmp_to_jpeg(wClipboard* clipboard, UINT32 formatId,
-                                                    const void* data, UINT32* pSize)
-{
-	return clipboard_synthesize_image_bmp_to_format(clipboard, formatId, WINPR_IMAGE_JPEG, data,
-	                                                pSize);
-}
-
-WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_image_jpeg_to_bmp(wClipboard* clipboard, UINT32 formatId,
-                                                    const void* data, UINT32* pSize)
-{
-	return clipboard_synthesize_image_format_to_bmp(clipboard, formatId, data, pSize);
+	free(dst);
+	return result;
 }
-#endif
 
 /**
  * "HTML Format":
@@ -769,9 +615,27 @@ static void* clipboard_synthesize_image_jpeg_to_bmp(wClipboard* clipboard, UINT3
  * HTML clipboard format: msdn.microsoft.com/en-us/library/windows/desktop/ms649015/
  */
 WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_html_format(wClipboard* clipboard, UINT32 formatId,
-                                              const void* pData, UINT32* pSize)
+static void* clipboard_synthesize_ms_html_format(wClipboard* clipboard, UINT32 formatId,
+                                                 const void* pData, UINT32* pSize)
 {
+	if (formatId != ClipboardGetFormatId(clipboard, mime_ms_html))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+	if (clipboard->formatId != ClipboardGetFormatId(clipboard, mime_html))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
 	union
 	{
 		const void* cpv;
@@ -987,7 +851,7 @@ static void* clipboard_wrap_format_to_html(uint32_t bmpFormat, const char* idata
 }
 
 WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_wrap_bmp_to_html(const char* idata, size_t ilength, uint32_t* plen)
+static void* clipboard_wrap_bmp_to_ms_html(const char* idata, size_t ilength, uint32_t* plen)
 {
 	const uint32_t formats[] = { WINPR_IMAGE_WEBP, WINPR_IMAGE_PNG, WINPR_IMAGE_JPEG };
 
@@ -1004,14 +868,24 @@ static void* clipboard_wrap_bmp_to_html(const char* idata, size_t ilength, uint3
 }
 
 WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_image_html(WINPR_ATTR_UNUSED wClipboard* clipboard,
-                                             UINT32 formatId, const void* data, UINT32* pSize)
+static void* clipboard_synthesize_image_ms_html(WINPR_ATTR_UNUSED wClipboard* clipboard,
+                                                UINT32 formatId, const void* data, UINT32* pSize)
 {
+	if (formatId != ClipboardGetFormatId(clipboard, mime_ms_html))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+
 	WINPR_ASSERT(pSize);
 
 	const size_t datalen = *pSize;
 
-	switch (formatId)
+	switch (clipboard->formatId)
 	{
 		case CF_TIFF:
 			return clipboard_wrap_html(mime_tiff, data, datalen, pSize);
@@ -1019,55 +893,55 @@ static void* clipboard_synthesize_image_html(WINPR_ATTR_UNUSED wClipboard* clipb
 		case CF_DIBV5:
 		{
 			uint32_t bmplen = *pSize;
-			void* bmp = clipboard_synthesize_image_bmp(clipboard, formatId, data, &bmplen);
+			void* bmp =
+			    clipboard_synthesize_image_bmp(clipboard, clipboard->formatId, data, &bmplen);
 			if (!bmp)
 			{
-				WLog_WARN(TAG, "failed to convert formatId 0x%08" PRIx32 " [%s]", formatId,
-				          ClipboardGetFormatName(clipboard, formatId));
+				WLog_WARN(TAG, "failed to convert formatId 0x%08" PRIx32 " [%s]",
+				          clipboard->formatId,
+				          ClipboardGetFormatName(clipboard, clipboard->formatId));
 				*pSize = 0;
 				return nullptr;
 			}
 
-			void* res = clipboard_wrap_bmp_to_html(bmp, bmplen, pSize);
+			void* res = clipboard_wrap_bmp_to_ms_html(bmp, bmplen, pSize);
 			free(bmp);
 			return res;
 		}
 		default:
 		{
+#if defined(WINPR_UTILS_IMAGE_WEBP)
 			const uint32_t idWebp = ClipboardRegisterFormat(clipboard, mime_webp);
-			const uint32_t idPng = ClipboardRegisterFormat(clipboard, mime_png);
-			const uint32_t idJpeg = ClipboardRegisterFormat(clipboard, mime_jpeg);
-			const uint32_t idTiff = ClipboardRegisterFormat(clipboard, mime_tiff);
-			if (formatId == idWebp)
-			{
+			if (clipboard->formatId == idWebp)
 				return clipboard_wrap_html(mime_webp, data, datalen, pSize);
-			}
-			else if (formatId == idPng)
-			{
+#endif
+
+#if defined(WINPR_UTILS_IMAGE_PNG)
+			const uint32_t idPng = ClipboardRegisterFormat(clipboard, mime_png);
+			if (clipboard->formatId == idPng)
 				return clipboard_wrap_html(mime_png, data, datalen, pSize);
-			}
-			else if (formatId == idJpeg)
-			{
+#endif
+#if defined(WINPR_UTILS_IMAGE_JPEG)
+			const uint32_t idJpeg = ClipboardRegisterFormat(clipboard, mime_jpeg);
+			if (clipboard->formatId == idJpeg)
 				return clipboard_wrap_html(mime_jpeg, data, datalen, pSize);
-			}
-			else if (formatId == idTiff)
-			{
+#endif
+
+			const uint32_t idTiff = ClipboardRegisterFormat(clipboard, mime_tiff);
+			if (clipboard->formatId == idTiff)
 				return clipboard_wrap_html(mime_tiff, data, datalen, pSize);
-			}
-			else
+
+			for (size_t x = 0; x < ARRAYSIZE(mime_bitmap); x++)
 			{
-				for (size_t x = 0; x < ARRAYSIZE(mime_bitmap); x++)
-				{
-					const char* mime = mime_bitmap[x];
-					const uint32_t id = ClipboardRegisterFormat(clipboard, mime);
+				const char* mime = mime_bitmap[x];
+				const uint32_t id = ClipboardRegisterFormat(clipboard, mime);
 
-					if (formatId == id)
-						return clipboard_wrap_bmp_to_html(data, datalen, pSize);
-				}
+				if (formatId == id)
+					return clipboard_wrap_bmp_to_ms_html(data, datalen, pSize);
 			}
 
-			WLog_WARN(TAG, "Unsupported image format id 0x%08" PRIx32 " [%s]", formatId,
-			          ClipboardGetFormatName(clipboard, formatId));
+			WLog_WARN(TAG, "Unsupported image format id 0x%08" PRIx32 " [%s]", clipboard->formatId,
+			          ClipboardGetFormatName(clipboard, clipboard->formatId));
 			*pSize = 0;
 			return nullptr;
 		}
@@ -1080,9 +954,28 @@ static void* clipboard_synthesize_image_html(WINPR_ATTR_UNUSED wClipboard* clipb
  * HTML text format.
  */
 WINPR_ATTR_MALLOC(free, 1)
-static void* clipboard_synthesize_text_html(wClipboard* clipboard, UINT32 formatId,
-                                            const void* data, UINT32* pSize)
+static void* clipboard_synthesize_html_format(wClipboard* clipboard, UINT32 formatId,
+                                              const void* data, UINT32* pSize)
 {
+	if (formatId != ClipboardGetFormatId(clipboard, mime_html))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+	if (clipboard->formatId != ClipboardGetFormatId(clipboard, mime_ms_html))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+
 	char* pDstData = nullptr;
 
 	if (formatId == ClipboardGetFormatId(clipboard, mime_ms_html))
@@ -1124,304 +1017,101 @@ static void* clipboard_synthesize_text_html(wClipboard* clipboard, UINT32 format
 
 BOOL ClipboardInitSynthesizers(wClipboard* clipboard)
 {
+	WINPR_ASSERT(clipboard);
+
+	const UINT32 formatIdUtf = ClipboardRegisterFormat(clipboard, mime_text_utf8);
+	const UINT32 formatIdUtf8String = ClipboardRegisterFormat(clipboard, mime_text_UTF8_STRING);
+	const UINT32 formatIdPlain = ClipboardRegisterFormat(clipboard, mime_text_plain);
+	const UINT32 textFormatIds[] = { CF_TEXT,     CF_OEMTEXT,         CF_UNICODETEXT,
+		                             formatIdUtf, formatIdUtf8String, formatIdPlain };
 	/**
 	 * CF_TEXT
 	 */
+	for (size_t x = 0; x < ARRAYSIZE(textFormatIds); x++)
 	{
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_TEXT, CF_OEMTEXT,
-		                                  clipboard_synthesize_cf_oemtext))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_TEXT, CF_UNICODETEXT,
-		                                  clipboard_synthesize_cf_unicodetext))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_TEXT, CF_LOCALE,
-		                                  clipboard_synthesize_cf_locale))
-			return FALSE;
-		{
-			UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_text_plain);
-			if (!ClipboardRegisterSynthesizer(clipboard, CF_TEXT, altFormatId,
-			                                  clipboard_synthesize_utf8_string))
-				return FALSE;
-		}
-		{
-			UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_text_utf8);
-			if (!ClipboardRegisterSynthesizer(clipboard, CF_TEXT, altFormatId,
-			                                  clipboard_synthesize_utf8_string))
-				return FALSE;
-		}
-	}
-	/**
-	 * CF_OEMTEXT
-	 */
-	{
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_OEMTEXT, CF_TEXT,
-		                                  clipboard_synthesize_cf_text))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_OEMTEXT, CF_UNICODETEXT,
-		                                  clipboard_synthesize_cf_unicodetext))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_OEMTEXT, CF_LOCALE,
-		                                  clipboard_synthesize_cf_locale))
-			return FALSE;
-		{
-			UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_text_plain);
-			if (!ClipboardRegisterSynthesizer(clipboard, CF_OEMTEXT, altFormatId,
-			                                  clipboard_synthesize_utf8_string))
-				return FALSE;
-		}
-		{
-			UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_text_utf8);
-			if (!ClipboardRegisterSynthesizer(clipboard, CF_OEMTEXT, altFormatId,
-			                                  clipboard_synthesize_utf8_string))
-				return FALSE;
-		}
-	}
-	/**
-	 * CF_UNICODETEXT
-	 */
-	{
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_UNICODETEXT, CF_TEXT,
-		                                  clipboard_synthesize_cf_text))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_UNICODETEXT, CF_OEMTEXT,
-		                                  clipboard_synthesize_cf_oemtext))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_UNICODETEXT, CF_LOCALE,
-		                                  clipboard_synthesize_cf_locale))
-			return FALSE;
-		{
-			UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_text_plain);
-			if (!ClipboardRegisterSynthesizer(clipboard, CF_UNICODETEXT, altFormatId,
-			                                  clipboard_synthesize_utf8_string))
-				return FALSE;
-		}
-		{
-			UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_text_utf8);
-			if (!ClipboardRegisterSynthesizer(clipboard, CF_UNICODETEXT, altFormatId,
-			                                  clipboard_synthesize_utf8_string))
-				return FALSE;
-		}
-	}
-	/**
-	 * UTF8_STRING
-	 * text/plain;charset=urf-8
-	 */
-	{
-		UINT32 formatId = ClipboardRegisterFormat(clipboard, mime_text_utf8);
+		const UINT32 formatId = textFormatIds[x];
 
-		if (formatId)
+		for (size_t y = 0; y < ARRAYSIZE(textFormatIds); y++)
 		{
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, CF_TEXT,
-			                                  clipboard_synthesize_cf_text))
-				return FALSE;
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, CF_OEMTEXT,
-			                                  clipboard_synthesize_cf_oemtext))
-				return FALSE;
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, CF_UNICODETEXT,
-			                                  clipboard_synthesize_cf_unicodetext))
-				return FALSE;
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, CF_LOCALE,
-			                                  clipboard_synthesize_cf_locale))
-				return FALSE;
-		}
-	}
-	/**
-	 * text/plain
-	 */
-	{
-		UINT32 formatId = ClipboardRegisterFormat(clipboard, mime_text_plain);
+			const UINT32 dstFormatId = textFormatIds[y];
+			if (formatId == dstFormatId)
+				continue;
 
-		if (formatId)
-		{
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, CF_TEXT,
-			                                  clipboard_synthesize_cf_text))
-				return FALSE;
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, CF_OEMTEXT,
-			                                  clipboard_synthesize_cf_oemtext))
-				return FALSE;
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, CF_UNICODETEXT,
-			                                  clipboard_synthesize_cf_unicodetext))
-				return FALSE;
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, CF_LOCALE,
-			                                  clipboard_synthesize_cf_locale))
+			if (!ClipboardRegisterSynthesizerEx(clipboard, formatId, dstFormatId,
+			                                    clipboard_synthesize_string))
 				return FALSE;
 		}
+
+		if (!ClipboardRegisterSynthesizerEx(clipboard, formatId, CF_LOCALE,
+		                                    clipboard_synthesize_cf_locale))
+			return FALSE;
 	}
 
-	const uint32_t htmlFormat = ClipboardRegisterFormat(clipboard, mime_ms_html);
-	const uint32_t tiffFormat = ClipboardRegisterFormat(clipboard, mime_tiff);
+	const uint32_t msHtmlFormat = ClipboardRegisterFormat(clipboard, mime_ms_html);
 
 	/**
 	 * CF_TIFF
 	 */
-	if (!ClipboardRegisterSynthesizer(clipboard, CF_TIFF, htmlFormat,
-	                                  clipboard_synthesize_image_html))
-		return FALSE;
-	if (!ClipboardRegisterSynthesizer(clipboard, tiffFormat, htmlFormat,
-	                                  clipboard_synthesize_image_html))
+	if (!ClipboardRegisterSynthesizerEx(clipboard, CF_TIFF, msHtmlFormat,
+	                                    clipboard_synthesize_image_ms_html))
 		return FALSE;
 
 	/**
-	 * CF_DIB
+	 * CF_DIB / CF_DIBv5
 	 */
-	{
-#if defined(WINPR_UTILS_IMAGE_DIBv5)
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIB, CF_DIBV5,
-		                                  clipboard_synthesize_cf_dibv5))
-			return FALSE;
-#endif
-		for (size_t x = 0; x < ARRAYSIZE(mime_bitmap); x++)
-		{
-			const char* mime = mime_bitmap[x];
-			const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime);
-			if (altFormatId == 0)
-				continue;
-			if (!ClipboardRegisterSynthesizer(clipboard, CF_DIB, altFormatId,
-			                                  clipboard_synthesize_image_bmp))
-				return FALSE;
-		}
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIB, htmlFormat,
-		                                  clipboard_synthesize_image_html))
-			return FALSE;
-	}
+	if (!ClipboardRegisterSynthesizerEx(clipboard, CF_DIBV5, CF_DIB,
+	                                    clipboard_synthesize_image_dib_to_format))
+		return FALSE;
+	if (!ClipboardRegisterSynthesizerEx(clipboard, CF_DIB, msHtmlFormat,
+	                                    clipboard_synthesize_image_ms_html))
+		return FALSE;
 
-	/**
-	 * CF_DIBV5
-	 */
 #if defined(WINPR_UTILS_IMAGE_DIBv5)
-	{
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIBV5, CF_DIB, clipboard_synthesize_cf_dib))
-			return FALSE;
-
-		for (size_t x = 0; x < ARRAYSIZE(mime_bitmap); x++)
-		{
-			const char* mime = mime_bitmap[x];
-			const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime);
-			if (altFormatId == 0)
-				continue;
-			if (!ClipboardRegisterSynthesizer(clipboard, CF_DIBV5, altFormatId,
-			                                  clipboard_synthesize_image_bmp))
-				return FALSE;
-		}
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIBV5, htmlFormat,
-		                                  clipboard_synthesize_image_html))
-			return FALSE;
-	}
+	if (!ClipboardRegisterSynthesizerEx(clipboard, CF_DIBV5, CF_DIB,
+	                                    clipboard_synthesize_image_dib_to_format))
+		return FALSE;
+	if (!ClipboardRegisterSynthesizerEx(clipboard, CF_DIBV5, msHtmlFormat,
+	                                    clipboard_synthesize_image_ms_html))
+		return FALSE;
 #endif
 
 	/**
-	 * image/bmp
+	 * image/
 	 */
-	for (size_t x = 0; x < ARRAYSIZE(mime_bitmap); x++)
+	for (size_t x = 0; x < ARRAYSIZE(mime_images); x++)
 	{
-		const char* mime = mime_bitmap[x];
+		const char* mime = mime_images[x];
 		const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime);
 		if (altFormatId == 0)
 			continue;
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, CF_DIB,
-		                                  clipboard_synthesize_cf_dib))
-			return FALSE;
-#if defined(WINPR_UTILS_IMAGE_DIBv5)
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, CF_DIBV5,
-		                                  clipboard_synthesize_cf_dibv5))
-			return FALSE;
-#endif
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, htmlFormat,
-		                                  clipboard_synthesize_image_html))
-			return FALSE;
-	}
-
-	/**
-	 * image/png
-	 */
-#if defined(WINPR_UTILS_IMAGE_PNG)
-	{
-		const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_png);
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIB, altFormatId,
-		                                  clipboard_synthesize_image_bmp_to_png))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, CF_DIB,
-		                                  clipboard_synthesize_image_png_to_bmp))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, htmlFormat,
-		                                  clipboard_synthesize_image_html))
-			return FALSE;
-#if defined(WINPR_UTILS_IMAGE_DIBv5)
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIBV5, altFormatId,
-		                                  clipboard_synthesize_image_bmp_to_png))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, CF_DIBV5,
-		                                  clipboard_synthesize_image_png_to_bmp))
-			return FALSE;
-#endif
-	}
-#endif
-
-	/**
-	 * image/webp
-	 */
-#if defined(WINPR_UTILS_IMAGE_WEBP)
-	{
-		const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_webp);
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIB, altFormatId,
-		                                  clipboard_synthesize_image_bmp_to_webp))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, CF_DIB,
-		                                  clipboard_synthesize_image_webp_to_bmp))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, htmlFormat,
-		                                  clipboard_synthesize_image_html))
-			return FALSE;
-#if defined(WINPR_UTILS_IMAGE_DIBv5)
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIBV5, altFormatId,
-		                                  clipboard_synthesize_image_bmp_to_webp))
-			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, CF_DIBV5,
-		                                  clipboard_synthesize_image_webp_to_bmp))
-			return FALSE;
-#endif
-	}
-#endif
-
-	/**
-	 * image/jpeg
-	 */
-#if defined(WINPR_UTILS_IMAGE_JPEG)
-	{
-		const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_jpeg);
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIB, altFormatId,
-		                                  clipboard_synthesize_image_bmp_to_jpeg))
+		if (!ClipboardRegisterSynthesizerEx(clipboard, CF_DIB, altFormatId,
+		                                    clipboard_synthesize_image_dib_to_format))
 			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, CF_DIB,
-		                                  clipboard_synthesize_image_jpeg_to_bmp))
+		if (!ClipboardRegisterSynthesizerEx(clipboard, altFormatId, CF_DIB,
+		                                    clipboard_synthesize_image_format_to_cf_dib))
 			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, htmlFormat,
-		                                  clipboard_synthesize_image_html))
+		if (!ClipboardRegisterSynthesizerEx(clipboard, altFormatId, msHtmlFormat,
+		                                    clipboard_synthesize_image_ms_html))
 			return FALSE;
 #if defined(WINPR_UTILS_IMAGE_DIBv5)
-		if (!ClipboardRegisterSynthesizer(clipboard, altFormatId, CF_DIBV5,
-		                                  clipboard_synthesize_image_jpeg_to_bmp))
+		if (!ClipboardRegisterSynthesizerEx(clipboard, CF_DIBV5, altFormatId,
+		                                    clipboard_synthesize_image_dib_to_format))
 			return FALSE;
-		if (!ClipboardRegisterSynthesizer(clipboard, CF_DIBV5, altFormatId,
-		                                  clipboard_synthesize_image_bmp_to_jpeg))
+		if (!ClipboardRegisterSynthesizerEx(clipboard, altFormatId, CF_DIBV5,
+		                                    clipboard_synthesize_image_format_to_cf_dib))
 			return FALSE;
 #endif
 	}
-#endif
 
 	/**
 	 * HTML Format
 	 */
+	if (msHtmlFormat)
 	{
-		UINT32 formatId = ClipboardRegisterFormat(clipboard, mime_ms_html);
-
-		if (formatId)
-		{
-			const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_html);
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, altFormatId,
-			                                  clipboard_synthesize_text_html))
-				return FALSE;
-		}
+		const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_html);
+		if (!ClipboardRegisterSynthesizerEx(clipboard, msHtmlFormat, altFormatId,
+		                                    clipboard_synthesize_html_format))
+			return FALSE;
 	}
 
 	/**
@@ -1433,8 +1123,8 @@ BOOL ClipboardInitSynthesizers(wClipboard* clipboard)
 		if (formatId)
 		{
 			const UINT32 altFormatId = ClipboardRegisterFormat(clipboard, mime_ms_html);
-			if (!ClipboardRegisterSynthesizer(clipboard, formatId, altFormatId,
-			                                  clipboard_synthesize_html_format))
+			if (!ClipboardRegisterSynthesizerEx(clipboard, formatId, altFormatId,
+			                                    clipboard_synthesize_ms_html_format))
 				return FALSE;
 		}
 	}
diff --git a/winpr/libwinpr/clipboard/synthetic_file.c b/winpr/libwinpr/clipboard/synthetic_file.c
index 86b55fdac44e..97f87f102fc1 100644
--- a/winpr/libwinpr/clipboard/synthetic_file.c
+++ b/winpr/libwinpr/clipboard/synthetic_file.c
@@ -563,12 +563,27 @@ static void* convert_any_uri_list_to_filedescriptors(wClipboard* clipboard,
 static void* convert_uri_list_to_filedescriptors(wClipboard* clipboard, UINT32 formatId,
                                                  const void* data, UINT32* pSize)
 {
-	const UINT32 expected = ClipboardGetFormatId(clipboard, mime_uri_list);
-	if (formatId != expected)
+	if (formatId != ClipboardGetFormatId(clipboard, mime_FileGroupDescriptorW))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+	if (clipboard->formatId != ClipboardGetFormatId(clipboard, mime_uri_list))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
 		return nullptr;
+	}
 	if (!process_uri_list(clipboard, (const char*)data, *pSize))
 		return nullptr;
-	return convert_any_uri_list_to_filedescriptors(clipboard, formatId, pSize);
+	return convert_any_uri_list_to_filedescriptors(clipboard, clipboard->formatId, pSize);
 }
 
 static BOOL process_files(wClipboard* clipboard, const char* data, UINT32 pSize, const char* prefix)
@@ -630,20 +645,50 @@ static BOOL process_mate_copied_files(wClipboard* clipboard, const char* data, U
 static void* convert_gnome_copied_files_to_filedescriptors(wClipboard* clipboard, UINT32 formatId,
                                                            const void* data, UINT32* pSize)
 {
-	const UINT32 expected = ClipboardGetFormatId(clipboard, mime_gnome_copied_files);
-	if (formatId != expected)
+	if (formatId != ClipboardGetFormatId(clipboard, mime_FileGroupDescriptorW))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+	if (clipboard->formatId != ClipboardGetFormatId(clipboard, mime_gnome_copied_files))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
 		return nullptr;
+	}
 	if (!process_gnome_copied_files(clipboard, (const char*)data, *pSize))
 		return nullptr;
-	return convert_any_uri_list_to_filedescriptors(clipboard, formatId, pSize);
+	return convert_any_uri_list_to_filedescriptors(clipboard, clipboard->formatId, pSize);
 }
 
 static void* convert_mate_copied_files_to_filedescriptors(wClipboard* clipboard, UINT32 formatId,
                                                           const void* data, UINT32* pSize)
 {
-	const UINT32 expected = ClipboardGetFormatId(clipboard, mime_mate_copied_files);
-	if (formatId != expected)
+	if (formatId != ClipboardGetFormatId(clipboard, mime_FileGroupDescriptorW))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
 		return nullptr;
+	}
+	if (clipboard->formatId != ClipboardGetFormatId(clipboard, mime_mate_copied_files))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
 
 	if (!process_mate_copied_files(clipboard, (const char*)data, *pSize))
 		return nullptr;
@@ -871,6 +916,24 @@ WINPR_ATTR_MALLOC(free, 1)
 static void* convert_filedescriptors_to_uri_list(wClipboard* clipboard, UINT32 formatId,
                                                  const void* data, UINT32* pSize)
 {
+	if (formatId != ClipboardGetFormatId(clipboard, mime_uri_list))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+	if (clipboard->formatId != ClipboardGetFormatId(clipboard, mime_FileGroupDescriptorW))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
 	return convert_filedescriptors_to_file_list(clipboard, formatId, data, pSize, "", "file://",
 	                                            "\r\n", FALSE);
 }
@@ -880,6 +943,24 @@ WINPR_ATTR_MALLOC(free, 1)
 static void* convert_filedescriptors_to_gnome_copied_files(wClipboard* clipboard, UINT32 formatId,
                                                            const void* data, UINT32* pSize)
 {
+	if (formatId != ClipboardGetFormatId(clipboard, mime_gnome_copied_files))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+	if (clipboard->formatId != ClipboardGetFormatId(clipboard, mime_FileGroupDescriptorW))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
 	return convert_filedescriptors_to_file_list(clipboard, formatId, data, pSize, "copy\n",
 	                                            "file://", "\n", TRUE);
 }
@@ -888,7 +969,24 @@ WINPR_ATTR_MALLOC(free, 1)
 static void* convert_filedescriptors_to_mate_copied_files(wClipboard* clipboard, UINT32 formatId,
                                                           const void* data, UINT32* pSize)
 {
-
+	if (formatId != ClipboardGetFormatId(clipboard, mime_mate_copied_files))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
+	if (clipboard->formatId != ClipboardGetFormatId(clipboard, mime_FileGroupDescriptorW))
+	{
+		WLog_ERR(TAG,
+		         "Unuspported destination format %s [0x%04" PRIx32
+		         "], trying to convert from %s [0x%04" PRIx32 "]",
+		         ClipboardGetFormatName(clipboard, formatId), formatId,
+		         ClipboardGetFormatName(clipboard, clipboard->formatId), clipboard->formatId);
+		return nullptr;
+	}
 	char* pDstData = convert_filedescriptors_to_file_list(clipboard, formatId, data, pSize,
 	                                                      "copy\n", "file://", "\n", TRUE);
 	if (!pDstData)
@@ -956,28 +1054,28 @@ static BOOL register_file_formats_and_synthesizers(wClipboard* clipboard)
 	obj = ArrayList_Object(clipboard->localFiles);
 	obj->fnObjectFree = array_free_synthetic_file;
 
-	if (!ClipboardRegisterSynthesizer(clipboard, local_file_format_id, file_group_format_id,
-	                                  convert_uri_list_to_filedescriptors))
+	if (!ClipboardRegisterSynthesizerEx(clipboard, local_file_format_id, file_group_format_id,
+	                                    convert_uri_list_to_filedescriptors))
 		goto error_free_local_files;
 
-	if (!ClipboardRegisterSynthesizer(clipboard, file_group_format_id, local_file_format_id,
-	                                  convert_filedescriptors_to_uri_list))
+	if (!ClipboardRegisterSynthesizerEx(clipboard, file_group_format_id, local_file_format_id,
+	                                    convert_filedescriptors_to_uri_list))
 		goto error_free_local_files;
 
-	if (!ClipboardRegisterSynthesizer(clipboard, local_gnome_file_format_id, file_group_format_id,
-	                                  convert_gnome_copied_files_to_filedescriptors))
+	if (!ClipboardRegisterSynthesizerEx(clipboard, local_gnome_file_format_id, file_group_format_id,
+	                                    convert_gnome_copied_files_to_filedescriptors))
 		goto error_free_local_files;
 
-	if (!ClipboardRegisterSynthesizer(clipboard, file_group_format_id, local_gnome_file_format_id,
-	                                  convert_filedescriptors_to_gnome_copied_files))
+	if (!ClipboardRegisterSynthesizerEx(clipboard, file_group_format_id, local_gnome_file_format_id,
+	                                    convert_filedescriptors_to_gnome_copied_files))
 		goto error_free_local_files;
 
-	if (!ClipboardRegisterSynthesizer(clipboard, local_mate_file_format_id, file_group_format_id,
-	                                  convert_mate_copied_files_to_filedescriptors))
+	if (!ClipboardRegisterSynthesizerEx(clipboard, local_mate_file_format_id, file_group_format_id,
+	                                    convert_mate_copied_files_to_filedescriptors))
 		goto error_free_local_files;
 
-	if (!ClipboardRegisterSynthesizer(clipboard, file_group_format_id, local_mate_file_format_id,
-	                                  convert_filedescriptors_to_mate_copied_files))
+	if (!ClipboardRegisterSynthesizerEx(clipboard, file_group_format_id, local_mate_file_format_id,
+	                                    convert_filedescriptors_to_mate_copied_files))
 		goto error_free_local_files;
 
 	return TRUE;
diff --git a/winpr/libwinpr/clipboard/test/TestClipboardFormats.c b/winpr/libwinpr/clipboard/test/TestClipboardFormats.c
index cfa0e2b1f946..ec0217fa9046 100644
--- a/winpr/libwinpr/clipboard/test/TestClipboardFormats.c
+++ b/winpr/libwinpr/clipboard/test/TestClipboardFormats.c
@@ -6,42 +6,90 @@
 #include <winpr/stream.h>
 #include <winpr/user.h>
 
+#define test_log(fmt, ...) test_log_fn(__FILE__, __func__, __LINE__, (fmt), ##__VA_ARGS__)
+static void test_log_fn(const char* file, const char* func, size_t line, const char* fmt, ...)
+{
+	va_list ap;
+	va_start(ap, fmt);
+
+	char* fstr = nullptr;
+	size_t fslen = 0;
+	winpr_vasprintf(&fstr, &fslen, fmt, ap);
+	(void)fprintf(stderr, "[%" PRIu32 ":%s]: %s\n", line, func, fstr);
+	free(fstr);
+	va_end(ap);
+}
+
+#define test_ClipboardSetData(clipboard, formatId, data, size) \
+	test_ClipboardSetDataFn(__FILE__, __func__, __LINE__, (clipboard), (formatId), (data), (size))
+static BOOL test_ClipboardSetDataFn(const char* file, const char* func, size_t line,
+                                    wClipboard* clipboard, UINT32 formatId, const void* data,
+                                    UINT32 size)
+{
+	const BOOL rc = ClipboardSetData(clipboard, formatId, data, size);
+	if (!rc)
+	{
+		test_log_fn(file, func, line, "ClipboardSetData %s[0x%04" PRIx32 "][%" PRIu32 "] failed",
+		            ClipboardGetFormatName(clipboard, formatId), formatId, size);
+	}
+	return rc;
+}
+
+#define test_ClipboardGetData(clipboard, formatId, size) \
+	test_ClipboardGetDataFn(__FILE__, __func__, __LINE__, (clipboard), (formatId), (size))
+static void* test_ClipboardGetDataFn(const char* file, const char* func, size_t line,
+                                     wClipboard* clipboard, UINT32 formatId, UINT32* pSize)
+{
+	void* rc = ClipboardGetData(clipboard, formatId, pSize);
+	if (!rc)
+	{
+		test_log_fn(file, func, line, "ClipboardGetData %s[0x%04" PRIx32 "] failed",
+		            ClipboardGetFormatName(clipboard, formatId), formatId);
+	}
+	return rc;
+}
+
+WINPR_ATTR_NODISCARD
 static BOOL test_dib_to_bmp(const BYTE* dib, size_t dibSize, size_t expectedOffset)
 {
 	BOOL rc = FALSE;
 
 	wClipboard* clipboard = ClipboardCreate();
 	if (!clipboard)
+	{
+		test_log("ClipboardCreate failed");
 		return FALSE;
+	}
 
 	const UINT32 bmpId = ClipboardRegisterFormat(clipboard, "image/bmp");
-	if ((bmpId == 0) || (dibSize > UINT32_MAX) ||
-	    !ClipboardSetData(clipboard, CF_DIB, dib, (UINT32)dibSize))
+	if (bmpId == 0)
+	{
+		test_log("bmpId == 0");
+		goto fail;
+	}
+	if (dibSize > UINT32_MAX)
+	{
+		test_log("dibSize %" PRIuz " > UINT32_MAX failed");
+		goto fail;
+	}
+	if (!test_ClipboardSetData(clipboard, CF_DIB, dib, (UINT32)dibSize))
 		goto fail;
 
 	UINT32 bmpSize = 0;
-	BYTE* bmp = ClipboardGetData(clipboard, bmpId, &bmpSize);
+	BYTE* bmp = test_ClipboardGetData(clipboard, bmpId, &bmpSize);
 	if (!bmp)
 		goto fail;
 
-	wStream bmpBuffer = WINPR_C_ARRAY_INIT;
-	wStream* s = Stream_StaticConstInit(&bmpBuffer, bmp, bmpSize);
-	if (!s || (bmpSize < sizeof(WINPR_BITMAP_FILE_HEADER)))
-		goto fail_bmp;
-
-	Stream_Seek(s, 10);
-	UINT32 bitmapOffset = 0;
-	Stream_Read_UINT32(s, bitmapOffset);
-	if ((bitmapOffset != expectedOffset) ||
-	    (bmpSize != sizeof(WINPR_BITMAP_FILE_HEADER) + dibSize) ||
-	    (memcmp(&bmp[sizeof(WINPR_BITMAP_FILE_HEADER)], dib, dibSize) != 0))
-		goto fail_bmp;
-
 	/* The synthesized BMP must also be readable by the image conversion code. */
 	wImage* image = winpr_image_new();
 	if (!image)
+	{
+		test_log("winpr_image_new failed");
 		goto fail_bmp;
+	}
 	rc = (winpr_image_read_buffer(image, bmp, bmpSize) > 0);
+	if (!rc)
+		test_log("winpr_image_read_buffer failed");
 	winpr_image_free(image, TRUE);
 
 fail_bmp:
@@ -65,6 +113,7 @@ static void write_dib_info_header(wStream* s, UINT32 size, UINT16 bpp, UINT32 co
 	Stream_Zero(s, 16);        /* resolution and palette metadata */
 }
 
+WINPR_ATTR_NODISCARD
 static BOOL test_dib_offsets(void)
 {
 	BYTE v4[sizeof(BITMAPV4HEADER) + 4] = WINPR_C_ARRAY_INIT;
@@ -109,6 +158,7 @@ static BOOL test_dib_offsets(void)
 	                           256 * sizeof(RGBQUAD));
 }
 
+WINPR_ATTR_NODISCARD
 static BOOL test_text_leading_newline(void)
 {
 	const char* tests[][2] = { { "\nfoo", "\r\nfoo" }, { "\n", "\r\n" } };
@@ -116,24 +166,27 @@ static BOOL test_text_leading_newline(void)
 
 	wClipboard* clipboard = ClipboardCreate();
 	if (!clipboard)
+	{
+		test_log("ClipboardCreate failed");
 		return FALSE;
+	}
 
 	const UINT32 textId = ClipboardRegisterFormat(clipboard, "text/plain");
 	for (size_t x = 0; x < ARRAYSIZE(tests); x++)
 	{
 		/* local text/plain is set without a terminator */
-		if (!ClipboardSetData(clipboard, textId, tests[x][0], (UINT32)strlen(tests[x][0])))
+		if (!test_ClipboardSetData(clipboard, textId, tests[x][0], (UINT32)strlen(tests[x][0])))
 			goto fail;
 
 		UINT32 size = 0;
-		WCHAR* wstr = ClipboardGetData(clipboard, CF_UNICODETEXT, &size);
+		WCHAR* wstr = test_ClipboardGetData(clipboard, CF_UNICODETEXT, &size);
 		char* str = ConvertWCharNToUtf8Alloc(wstr, size / sizeof(WCHAR), nullptr);
 		const BOOL match = str && (strcmp(str, tests[x][1]) == 0);
 		free(wstr);
 		free(str);
 		if (!match)
 		{
-			(void)fprintf(stderr, "text/plain to CF_UNICODETEXT failed for case %" PRIuz "\n", x);
+			test_log("text/plain to CF_UNICODETEXT failed for case %" PRIuz "", x);
 			goto fail;
 		}
 	}
@@ -144,33 +197,115 @@ static BOOL test_text_leading_newline(void)
 	return rc;
 }
 
+WINPR_ATTR_NODISCARD
+static BOOL test_text_conversion(void)
+{
+	wClipboard* clipboard = ClipboardCreate();
+	if (!clipboard)
+		return FALSE;
+
+	typedef struct
+	{
+		UINT32 srcFormat;
+		UINT32 dstFormat;
+		const char* src;
+		size_t srcLen;
+		const char* expect;
+		size_t expectLen;
+	} test_case_t;
+
+	const test_case_t tests[] = {
+		{ ClipboardRegisterFormat(clipboard, "text/plain"),
+		  ClipboardRegisterFormat(clipboard, "text/plain;charset=utf-8"), "a\nb\rc\\u1234\\u055e",
+		  17, "a\nb\ncሴ՞", 10 },
+		{ ClipboardRegisterFormat(clipboard, "text/plain;charset=utf-8"),
+		  ClipboardRegisterFormat(clipboard, "text/plain"), "a\r\nb\nc՞՞", 10,
+		  "a\r\nb\nc\\u055e\\u055e", 18 },
+		{ ClipboardRegisterFormat(clipboard, "text/plain"), CF_UNICODETEXT, "a\nb\rc\\u1234\\u055e",
+		  17, "\x61\x00\x0d\x00\x0a\x00\x62\x00\x0d\x0\x0a\x00\x63\x00\x34\x12\x5e\x05\x00\x00\x00",
+		  18 },
+		{ CF_UNICODETEXT, ClipboardRegisterFormat(clipboard, "text/plain"),
+		  "\x61\x00\x0d\x00\x0a\x00\x62\x00\x0d\x0\x0a\x00\x63\x00\x34\x12\x5e\x05\x00\x00\x00", 18,
+		  "a\nb\nc\\u1234\\u055e", 17 },
+		{ ClipboardRegisterFormat(clipboard, "text/plain"), CF_TEXT, "a\nb\rc\\u1234\\u055e", 17,
+		  "a\nb\nc\\u1234\\u055e", 17 },
+		{ CF_TEXT, ClipboardRegisterFormat(clipboard, "text/plain"), "a\nb\nc\\u1234\\u055e", 17,
+		  "a\nb\nc\\u1234\\u055e", 17 },
+		{ ClipboardRegisterFormat(clipboard, "text/plain"), CF_OEMTEXT, "a\nb\rc\\u1234\\u055e", 17,
+		  "a\nb\nc\\u1234\\u055e", 17 },
+		{ CF_OEMTEXT, ClipboardRegisterFormat(clipboard, "text/plain"), "a\nb\rc\\u1234\\u055e", 17,
+		  "a\nb\nc\\u1234\\u055e", 17 }
+	};
+
+	BOOL rc = FALSE;
+
+	for (size_t x = 0; x < ARRAYSIZE(tests); x++)
+	{
+		const test_case_t* cur = &tests[x];
+
+		/* local text/plain is set without a terminator */
+		if (!test_ClipboardSetData(clipboard, cur->srcFormat, cur->src, cur->srcLen))
+			goto fail;
+
+		UINT32 dstSize = 0;
+		BYTE* data = test_ClipboardGetData(clipboard, cur->dstFormat, &dstSize);
+		if (dstSize != cur->expectLen)
+		{
+			winpr_znfree(data, dstSize);
+			goto fail;
+		}
+		if (dstSize != 0)
+		{
+			if (!data)
+			{
+				winpr_znfree(data, dstSize);
+				goto fail;
+			}
+			const int rc = memcmp(cur->expect, data, cur->expectLen);
+			winpr_znfree(data, dstSize);
+			if (rc != 0)
+				goto fail;
+		}
+		else
+			winpr_znfree(data, dstSize);
+	}
+	rc = TRUE;
+
+fail:
+	ClipboardDestroy(clipboard);
+	return rc;
+}
+
 int TestClipboardFormats(int argc, char* argv[])
 {
 	int rc = -1;
 	UINT32 count = 0;
 	UINT32* pFormatIds = nullptr;
 	const char* formatName = nullptr;
-	wClipboard* clipboard = nullptr;
 	UINT32 utf8StringFormatId = 0;
 
 	WINPR_UNUSED(argc);
 	WINPR_UNUSED(argv);
 
-	clipboard = ClipboardCreate();
+	if (!test_text_leading_newline())
+		return -1;
+
+	if (!test_text_conversion())
+		return -1;
+
+	wClipboard* clipboard = ClipboardCreate();
 	if (!clipboard)
 		return -1;
-	if (!test_dib_offsets())
-		goto fail;
-	if (!test_text_leading_newline())
-		goto fail;
 
-	const char* mime_types[] = { "text/html", "text/html",  "image/bmp",
-		                         "image/png", "image/webp", "image/jpeg" };
+	const char* mime_types[] = { "text/plain", "text/plain;charset=utf-8",
+		                         "text/html",  "image/bmp",
+		                         "image/png",  "image/webp",
+		                         "image/jpeg" };
 	for (size_t x = 0; x < ARRAYSIZE(mime_types); x++)
 	{
 		const char* mime = mime_types[x];
 		UINT32 id = ClipboardRegisterFormat(clipboard, mime);
-		(void)fprintf(stderr, "ClipboardRegisterFormat(%s) -> 0x%08" PRIx32 "\n", mime, id);
+		test_log("ClipboardRegisterFormat(%s) -> 0x%08" PRIx32 "", mime, id);
 		if (id == 0)
 			goto fail;
 	}
@@ -183,25 +318,26 @@ int TestClipboardFormats(int argc, char* argv[])
 	{
 		UINT32 formatId = pFormatIds[index];
 		formatName = ClipboardGetFormatName(clipboard, formatId);
-		(void)fprintf(stderr, "Format: 0x%08" PRIX32 " %s\n", formatId, formatName);
+		test_log("Format: 0x%08" PRIX32 " %s", formatId, formatName);
 	}
 
 	free(pFormatIds);
 
 	if (1)
 	{
-		BOOL bSuccess = 0;
 		UINT32 SrcSize = 0;
 		UINT32 DstSize = 0;
 		const char pSrcData[] = "this is a test string";
 		char* pDstData = nullptr;
 
 		SrcSize = (UINT32)(strnlen(pSrcData, ARRAYSIZE(pSrcData)) + 1);
-		bSuccess = ClipboardSetData(clipboard, utf8StringFormatId, pSrcData, SrcSize);
-		(void)fprintf(stderr, "ClipboardSetData: %" PRId32 "\n", bSuccess);
+		const BOOL bSuccess =
+		    test_ClipboardSetData(clipboard, utf8StringFormatId, pSrcData, SrcSize);
+		if (!bSuccess)
+			goto fail;
+
 		DstSize = 0;
-		pDstData = (char*)ClipboardGetData(clipboard, utf8StringFormatId, &DstSize);
-		(void)fprintf(stderr, "ClipboardGetData: %s\n", pDstData);
+		pDstData = (char*)test_ClipboardGetData(clipboard, utf8StringFormatId, &DstSize);
 		free(pDstData);
 	}
 
@@ -211,10 +347,9 @@ int TestClipboardFormats(int argc, char* argv[])
 		char* pSrcData = nullptr;
 		WCHAR* pDstData = nullptr;
 		DstSize = 0;
-		pDstData = (WCHAR*)ClipboardGetData(clipboard, CF_UNICODETEXT, &DstSize);
+		pDstData = (WCHAR*)test_ClipboardGetData(clipboard, CF_UNICODETEXT, &DstSize);
 		pSrcData = ConvertWCharNToUtf8Alloc(pDstData, DstSize / sizeof(WCHAR), nullptr);
 
-		(void)fprintf(stderr, "ClipboardGetData (synthetic): %s\n", pSrcData);
 		free(pDstData);
 		free(pSrcData);
 	}
@@ -226,7 +361,7 @@ int TestClipboardFormats(int argc, char* argv[])
 	{
 		UINT32 formatId = pFormatIds[index];
 		formatName = ClipboardGetFormatName(clipboard, formatId);
-		(void)fprintf(stderr, "Format: 0x%08" PRIX32 " %s\n", formatId, formatName);
+		test_log("Format: 0x%08" PRIX32 " %s", formatId, formatName);
 	}
 
 	if (1)
@@ -247,8 +382,7 @@ int TestClipboardFormats(int argc, char* argv[])
 
 		size_t bmpsize = 0;
 		void* data = winpr_image_write_buffer(img, WINPR_IMAGE_BITMAP, &bmpsize);
-		bSuccess = ClipboardSetData(clipboard, idBmp, data, bmpsize);
-		(void)fprintf(stderr, "ClipboardSetData: %" PRId32 "\n", bSuccess);
+		bSuccess = test_ClipboardSetData(clipboard, idBmp, data, bmpsize);
 
 		free(data);
 		winpr_image_free(img, TRUE);
@@ -259,12 +393,10 @@ int TestClipboardFormats(int argc, char* argv[])
 			UINT32 id = CF_DIB;
 
 			UINT32 DstSize = 0;
-			void* pDstData = ClipboardGetData(clipboard, id, &DstSize);
-			(void)fprintf(stderr, "ClipboardGetData: [CF_DIB] %p [%" PRIu32 "]\n", pDstData,
-			              DstSize);
+			void* pDstData = test_ClipboardGetData(clipboard, id, &DstSize);
 			if (!pDstData)
 				goto fail;
-			bSuccess = ClipboardSetData(clipboard, id, pDstData, DstSize);
+			bSuccess = test_ClipboardSetData(clipboard, id, pDstData, DstSize);
 			free(pDstData);
 			if (!bSuccess)
 				goto fail;
@@ -272,7 +404,7 @@ int TestClipboardFormats(int argc, char* argv[])
 		{
 			const uint32_t id = ClipboardGetFormatId(clipboard, "HTML Format");
 			UINT32 DstSize = 0;
-			void* pDstData = ClipboardGetData(clipboard, id, &DstSize);
+			void* pDstData = test_ClipboardGetData(clipboard, id, &DstSize);
 			if (!pDstData)
 				goto fail;
 			{
@@ -289,9 +421,7 @@ int TestClipboardFormats(int argc, char* argv[])
 			UINT32 id = ClipboardRegisterFormat(clipboard, "image/bmp");
 
 			UINT32 DstSize = 0;
-			void* pDstData = ClipboardGetData(clipboard, id, &DstSize);
-			(void)fprintf(stderr, "ClipboardGetData: [image/bmp] %p [%" PRIu32 "]\n", pDstData,
-			              DstSize);
+			void* pDstData = test_ClipboardGetData(clipboard, id, &DstSize);
 			if (!pDstData)
 				goto fail;
 			free(pDstData);
@@ -304,8 +434,7 @@ int TestClipboardFormats(int argc, char* argv[])
 			UINT32 id = ClipboardRegisterFormat(clipboard, "image/png");
 
 			UINT32 DstSize = 0;
-			void* pDstData = ClipboardGetData(clipboard, id, &DstSize);
-			(void)fprintf(stderr, "ClipboardGetData: [image/png] %p\n", pDstData);
+			void* pDstData = test_ClipboardGetData(clipboard, id, &DstSize);
 			if (!pDstData)
 				goto fail;
 			free(pDstData);
@@ -327,8 +456,7 @@ int TestClipboardFormats(int argc, char* argv[])
 
 			size_t bmpsize = 0;
 			void* data = winpr_image_write_buffer(img, WINPR_IMAGE_PNG, &bmpsize);
-			bSuccess = ClipboardSetData(clipboard, idBmp, data, bmpsize);
-			(void)fprintf(stderr, "ClipboardSetData: %" PRId32 "\n", bSuccess);
+			bSuccess = test_ClipboardSetData(clipboard, idBmp, data, bmpsize);
 
 			free(data);
 			winpr_image_free(img, TRUE);
@@ -339,12 +467,10 @@ int TestClipboardFormats(int argc, char* argv[])
 			UINT32 id = CF_DIB;
 
 			UINT32 DstSize = 0;
-			void* pDstData = ClipboardGetData(clipboard, id, &DstSize);
-			(void)fprintf(stderr, "ClipboardGetData: [CF_DIB] %p [%" PRIu32 "]\n", pDstData,
-			              DstSize);
+			void* pDstData = test_ClipboardGetData(clipboard, id, &DstSize);
 			if (!pDstData)
 				goto fail;
-			bSuccess = ClipboardSetData(clipboard, id, pDstData, DstSize);
+			bSuccess = test_ClipboardSetData(clipboard, id, pDstData, DstSize);
 			free(pDstData);
 			if (!bSuccess)
 				goto fail;
@@ -356,8 +482,7 @@ int TestClipboardFormats(int argc, char* argv[])
 			UINT32 id = ClipboardRegisterFormat(clipboard, "image/webp");
 
 			UINT32 DstSize = 0;
-			void* pDstData = ClipboardGetData(clipboard, id, &DstSize);
-			(void)fprintf(stderr, "ClipboardGetData: [image/webp] %p\n", pDstData);
+			void* pDstData = test_ClipboardGetData(clipboard, id, &DstSize);
 			if (!pDstData)
 				goto fail;
 			free(pDstData);
@@ -369,8 +494,7 @@ int TestClipboardFormats(int argc, char* argv[])
 			UINT32 id = ClipboardRegisterFormat(clipboard, "image/jpeg");
 
 			UINT32 DstSize = 0;
-			void* pDstData = ClipboardGetData(clipboard, id, &DstSize);
-			(void)fprintf(stderr, "ClipboardGetData: [image/jpeg] %p\n", pDstData);
+			void* pDstData = test_ClipboardGetData(clipboard, id, &DstSize);
 			if (!pDstData)
 				goto fail;
 			free(pDstData);
@@ -378,6 +502,9 @@ int TestClipboardFormats(int argc, char* argv[])
 #endif
 	}
 
+	if (!test_dib_offsets())
+		goto fail;
+
 	rc = 0;
 
 fail:
diff --git a/winpr/libwinpr/utils/image.c b/winpr/libwinpr/utils/image.c
index 4207be215fd6..33a2e106c805 100644
--- a/winpr/libwinpr/utils/image.c
+++ b/winpr/libwinpr/utils/image.c
@@ -240,9 +240,9 @@ BOOL readBitmapInfoHeader(wStream* s, WINPR_BITMAP_INFO_HEADER* bi, size_t* poff
 	return Stream_SafeSeek(s, bi->biSize - pos);
 }
 
-BYTE* winpr_bitmap_construct_header(size_t width, size_t height, size_t bpp)
+WINPR_ATTR_MALLOC(Stream_Free, 1)
+static wStream* winpr_bitmap_construct_header_stream(size_t width, size_t height, size_t bpp)
 {
-	BYTE* result = nullptr;
 	WINPR_BITMAP_FILE_HEADER bf = WINPR_C_ARRAY_INIT;
 	WINPR_BITMAP_INFO_HEADER bi = WINPR_C_ARRAY_INIT;
 
@@ -284,9 +284,29 @@ BYTE* winpr_bitmap_construct_header(size_t width, size_t height, size_t bpp)
 			{
 				DWORD used = bi.biClrUsed;
 				if (used == 0)
-					used = (1u << bi.biBitCount) / 8;
+					used = 1u << bi.biBitCount;
 				offset += sizeof(RGBQUAD) * used;
 			}
+			if (bi.biSizeImage == 0)
+			{
+				const UINT64 rawustride =
+				    WINPR_ASSERTING_INT_CAST(UINT64, bi.biWidth) * bi.biBitCount;
+				const UINT64 ustride = ((rawustride + 31) & ~31u) >> 3;
+				if (ustride > UINT32_MAX)
+				{
+					WLog_ERR(TAG, "bi->biWidth * bi->biBitCount > UINT32_MAX");
+					goto fail;
+				}
+
+				const UINT32 xstride = WINPR_ASSERTING_INT_CAST(uint32_t, ustride);
+				const UINT64 usize = WINPR_ASSERTING_INT_CAST(UINT64, llabs(bi.biHeight)) * xstride;
+				if (usize > UINT32_MAX)
+				{
+					WLog_ERR(TAG, "abs(bi->biHeight) * stride > UINT32_MAX");
+					goto fail;
+				}
+				bi.biSizeImage = WINPR_ASSERTING_INT_CAST(uint32_t, usize);
+			}
 			break;
 		case BI_BITFIELDS:
 			offset += sizeof(DWORD) * 3; // 3 DWORD color masks
@@ -295,6 +315,8 @@ BYTE* winpr_bitmap_construct_header(size_t width, size_t height, size_t bpp)
 			return nullptr;
 	}
 
+	bf.bfOffBits += offset;
+
 	if (!writeBitmapFileHeader(s, &bf))
 		goto fail;
 
@@ -308,10 +330,20 @@ BYTE* winpr_bitmap_construct_header(size_t width, size_t height, size_t bpp)
 	const size_t pos = Stream_GetPosition(s);
 	if (pos != bf.bfOffBits)
 		goto fail;
-	result = Stream_Buffer(s);
+	return s;
 fail:
-	Stream_Free(s, result == nullptr);
-	return result;
+	Stream_Free(s, TRUE);
+	return nullptr;
+}
+
+BYTE* winpr_bitmap_construct_header(size_t width, size_t height, size_t bpp)
+{
+	wStream* s = winpr_bitmap_construct_header_stream(width, height, bpp);
+	if (!s)
+		return nullptr;
+	void* data = Stream_Buffer(s);
+	Stream_Free(s, FALSE);
+	return data;
 }
 
 /**
@@ -356,13 +388,9 @@ static void* winpr_bitmap_write_buffer(const BYTE* data, size_t size, UINT32 wid
 	if (bpp_stride > stride)
 		return nullptr;
 
-	wStream* s = Stream_New(nullptr, 1024);
-	BYTE* bmp_header = winpr_bitmap_construct_header(width, height, bpp);
-	if (!bmp_header)
-		goto fail;
-	if (!Stream_EnsureRemainingCapacity(s, WINPR_IMAGE_BMP_HEADER_LEN))
+	wStream* s = winpr_bitmap_construct_header_stream(width, height, bpp);
+	if (!s)
 		goto fail;
-	Stream_Write(s, bmp_header, WINPR_IMAGE_BMP_HEADER_LEN);
 
 	for (size_t y = 0; y < height; y++)
 	{
@@ -389,7 +417,6 @@ static void* winpr_bitmap_write_buffer(const BYTE* data, size_t size, UINT32 wid
 	}
 fail:
 	Stream_Free(s, result == nullptr);
-	free(bmp_header);
 	return result;
 }
 
